src/main/kotlin/me/owdding/lib/cosmetics/CosmeticManager.kt

```diff
@@ -0,0 +1,152 @@
+import tech.thatgravyboat.skyblockapi.utils.json.Json.toDataOrThrow
+import java.net.URI
+import java.net.http.HttpClient
+import java.net.http.HttpRequest
+import java.net.http.HttpResponse
+import java.util.*
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.ConcurrentHashMap
+import kotlin.time.Duration.Companion.seconds
+import kotlin.time.toJavaDuration
+
+const val COSMETIC_VERSION = 1
+
+@Module
+object CosmeticManager {
+
+    private val _cosmetics: MutableMap<CosmeticId, Cosmetic> = ConcurrentHashMap()
+    val cosmetics: Map<CosmeticId, Cosmetic> get() = _cosmetics
+    val cosmeticList: Collection<Cosmetic> get() = _cosmetics.values
+
+    private val _players = ConcurrentHashMap<UUID, PlayerData>()
+    val players: Map<UUID, PlayerData> get() = _players
+    val playerList: Collection<PlayerData> get() = _players.values
+
+    private val _mlibCosmetics = ConcurrentHashMap<UUID, MlibCosmeticData>()
+    val mlibCosmetics: Map<UUID, MlibCosmeticData> get() = _mlibCosmetics
+
+    @JvmStatic
+    val imageProvider get() = CosmeticImageProvider
+
+    var client: HttpClient = HttpClient.newBuilder()
+        .connectTimeout(10.seconds.toJavaDuration())
+        .followRedirects(HttpClient.Redirect.ALWAYS)
+        .build()
+
+    init {
+        CompletableFuture.runAsync { loadData() }
+    }
+
+    fun loadData() {
+        _cosmetics.clear()
+        _players.clear()
+        val response = client.send(
+            HttpRequest.newBuilder(URI("https://cosmetics.meowdd.ing")).GET().build(),
+            HttpResponse.BodyHandlers.ofString(Charsets.UTF_8),
+        )
+        if (response.statusCode() != 200) return
+        val body = response.body().readJson<JsonObject>()
+        try {
+            body.toDataOrThrow<Unit>(
+                RecordCodecBuilder.create {
+                    it.group(
+                        MeowddingLibCodecs.CompletablePlayerDataCodec.codec().listOf().fieldOf("players").forGetter { emptyList() },
+                        MeowddingLibCodecs.CosmeticCodec.codec().listOf().fieldOf("cosmetics").forGetter { emptyList() },
+                    ).apply(it) { players, cosmetics ->
+                        val cosmetics = cosmetics.filter { (_, version) -> version >= COSMETIC_VERSION }
+                        this._cosmetics.putAll(cosmetics.associateBy(Cosmetic::id))
+                        this._players.putAll(players.map(CompletablePlayerData::complete).associateBy(PlayerData::uuid))
+                    }
+                },
+            )
+        } catch (e: Exception) {
+            e.printStackTrace()
+        }
+        this._mlibCosmetics.putAll(
+            playerList.associateNotNull(
+                keySelector = { it.uuid },
+                valueSelector = {
+                    it.data.toData(MeowddingLibCodecs.getCodec<MlibCosmeticData>())?.takeUnless { data -> data == emptyCosmetic }
+                },
+            ),
+        )
```

Comment:
the mlib use of the cosmetic stuff should be separated, this should instead fire an event that cosmetics were loaded and we can listen to it ourself for the built in ones